subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java

```diff
@@ -74,7 +75,8 @@ public ApiGroovyCompiler(Compiler<JavaCompileSpec> javaCompiler) {
 
     private static abstract class IncrementalCompilationCustomizer extends CompilationCustomizer {
         static IncrementalCompilationCustomizer fromSpec(GroovyJavaJointCompileSpec spec) {
-            if (spec.getCompilationMappingFile() != null) {
+            if (spec.getCompilationMappingFile() != null && !annotationProcessingConfigured(spec)) {
+                DeprecationLogger.nagUserOfDeprecatedBehaviour("Annotation processing is not supported when incremental Groovy compilation is enabled.");
```

Comment:
We can't do that in the `ApiGroovyCompiler`, we need to do it in `GroovyCompile`. This is running on the compiler daemon - I don't think we will see the deprecation message.

Reply:
PTAL!